### ISSUE TITLE
how to use an identical network to train molecule of different elements

### ISSUE THREAD

**Issue by tamaswells** [aiqm/torchani#304]
how can I use an identical network with species 'HCNO' to train molecule of element [CHO, CHNO, CO, HO and O]? For H2O, the species tensor is [0,0,3] and will be over the boundary of energy list.

**Comment by farhadrgh** [comment]
I would need more clarification about your question. Are you training using a particular dataset that only includes [CHO, CHNO, CO, HO, and O]? What do you mean by 'over the boundary of energy list'? What script are you using for training?

Here are a couple of general notes about training:
- If you have a dataset with the molecules you have mentioned above, you can follow the example file in [here](https://github.com/aiqm/torchani/blob/master/examples/nnp_training.py).
- The purpose of `energy_shifter` is to approximately shift the average of targets properties (molecular energies) to zero by finding Self Atomic Energies `SAE` through list square fitting.

**Comment by tamaswells** [comment]
Hi
I have modified your training script in order to utilize the trained network, where I still use the network framework containing "HCNO". Modifications are as follows:
1. I defined a function to convert element string to species torch.tensor by function "species_to_tensors"
2. I rewrote the class EnergyShifter since for H2O2, the species tensor is [[3,0,3,0]] while the energy self.self_energies in EnergyShifter.sae only contains two members. So I wrote a function fill_missing_energies to temporarily extend self.self_energies to a tensor containing four members. Besides, for molecule of only one specie, the dimension of self.self_energies is 0 and will make the operation self_energies = self.self_energies[species] crash.     
3. Is it proper to use best.pt rather than latest.pt for the real practice?   
  




  




------------------ 原始邮件 ------------------
发件人: "Farhad Ramezanghorbani"<notifications@github.com>;
发送时间: 2019年8月30日(星期五) 中午11:33
收件人: "aiqm/torchani"<torchani@noreply.github.com>;
抄送: "Xu Nan"<pplivegw@foxmail.com>;"Author"<author@noreply.github.com>;
主题: Re: [aiqm/torchani] how to use an identical network to train moleculeof different elements (#304)




I would need more clarification about your question. Are you training using a particular dataset that only includes [CHO, CHNO, CO, HO, and O]? What do you mean by 'over the boundary of energy list'? What script are you using for training?
 
Here are a couple of general notes about training:
  
If you have a dataset with the molecules you have mentioned above, you can follow the example file in here.
 
The purpose of energy_shifter is to approximately shift the average of targets properties (molecular energies) to zero by finding Self Atomic Energies SAE through list square fitting.

**Comment by tamaswells** [comment]
I am sorry the attached files of the replied email were missing. Could you give me your personal email address so I can seed you my files?  I intend to integrate torchANI into Lammps. Do you have any ideas for me?

**Comment by farhadrgh** [comment]
You can drag and drop the files on the bottom of comment session, or use GitHub Gist. My email address: farhad.ghorbani@gmail.com

The `best.pt` contains the best model trained. Therefore, it is the checkpoint you need to use for evaluating a trained model. `latest.pt` is the checkpoint used for resuming a training process.

> I am sorry the attached files of the replied email were missing. Could you give me your personal email address so I can seed you my files? I intend to integrate torchANI into Lammps. Do you have any ideas for me?

I would suggest taking a look at [ASE_ANI](https://github.com/isayev/ASE_ANI) interface which interfaces the ANI CUDA code (NeuroChem) with ASE.
